### PR TITLE
Initial contribution of EMC2101 plugin

### PR DESCRIPTION
**Pull request by JeremyLaurenson**
This plugin allows you to control the temperature of your printer enclosure using a 4 pin PWM fan connected to an Adafruit EMC2101 fan controller/temperature sensor. This Octoprint plugin is for use with a Raspberry Pi connected to an Adafruit EMC2101 fan control board.

### CODE REVIEW

**Review by jneilliii — CHANGES_REQUESTED** [review]
Along with the comments on this PR you may also want to add [this ](https://github.com/jneilliii/OctoPrint-GoogleDriveBackup/blob/337eafb067593ddba91bede3850373e00b803226/setup.py#L62) to your setup.py to avoid side-loading and people attempting to bypass the plugin manager for installs.

Instead of using a direct folder link it's probably better if you use self.get_plugin_data_folder() like [here](https://github.com/jneilliii/OctoPrint-Tasmota/blob/07913c6fa4fd0c63cb63e46537eac9fe2df77ffd/octoprint_tasmota/__init__.py#L114). This way the files will get saved with backup/restore and for people with multiple instances and separate basedir parameters starting OctoPrint. 

I'm not 100% sure that the way that you are using tempcontrol.py is the best solution, but may have to ask @cp2004 his thoughts on the matter. Of course it adds to the complexity a little because of the pathing change mentioned above.

File: _plugins/emc2101.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+description: Enable control of a PWM fan for your enclosue controlled by an Adafruit 2101
+authors:
+- Jeremy Laurenson
+license: AGPLv3
+
+# TODO
+date: 2022-02-14
+homepage: https://github.com/JeremyLaurenson/OctoPrint-Emc2101
+source: https://github.com/JeremyLaurenson/OctoPrint-Emc2101
+archive: https://github.com/JeremyLaurenson/OctoPrint-Emc2101/archive/master.zip
```
[the remark this answers is not in the document]
Comment (JeremyLaurenson) [reply _plugins/emc2101.md:15]: Thanks, done!

**Comment by JeremyLaurenson** [comment]
> Along with the comments on this PR you may also want to add [this ](https://github.com/jneilliii/OctoPrint-GoogleDriveBackup/blob/337eafb067593ddba91bede3850373e00b803226/setup.py#L62) to your setup.py to avoid side-loading and people attempting to bypass the plugin manager for installs.
 
Done.

> Instead of using a direct folder link it's probably better if you use self.get_plugin_data_folder() like [here](https://github.com/jneilliii/OctoPrint-Tasmota/blob/07913c6fa4fd0c63cb63e46537eac9fe2df77ffd/octoprint_tasmota/__init__.py#L114). This way the files will get saved with backup/restore and for people with multiple instances and separate basedir parameters starting OctoPrint.

Done

 
> I'm not 100% sure that the way that you are using tempcontrol.py is the best solution, but may have to ask @cp2004 his thoughts on the matter. Of course it adds to the complexity a little because of the pathing change mentioned above.

What I plan to do later is leave that python script running regardless of Octoprint since I want my enclosure cooled regardless. The EMC chip has a custom lookup table capability thats broken right now in the adafruit library and at that point I may only submit "changes" to the chip.

**Comment by cp2004** [comment]
Couple of my comments based on some poking around - @jneilliii is the review master, but since I was mentioned I took a peek... Addressing my comments are not necessary for the plugin to be on the repo, it is just (shall we say) additional.

I do agree that the way of running the script in the background will likely need some work. Will it prevent OctoPrint from shutting down if it is still running? Even if you don't see this behaviour, file handles can be inherited, serial ports can be locked and resources left in an unstable state if things aren't cleaned up. I've fought hard to debug issues with multiprocessing and non-daemon threads in the past blocking clean shutdown. I would love to be completely wrong on this, given I am just operating from feeling not actual testing/evidence.

If the script is continuing in the background, how does it work if someone restarts OctoPrint? Will you end up with running lots of versions of the script 

You are calling `python tempcontrol.py` (sort of) - this will usually be Python 2, sometimes not installed or available entirely (see Ubuntu 20.04, where `python` is no longer in the repository) or `Python 3`. If you are prepared to handle this, OK, but be aware of the potential problems. Lesson with 100s of thousands of installs of OctoPrint - people *will* do things you don't expect.

Personally, I would just run your Python script in a thread - it would cut down the code needed in that file to around 1/3 of it's current length, because you can keep the settings in one place. But at the end of the day, as long as you don't break anything in OctoPrint that people will come complaining about, it is in your control 😉.

https://github.com/JeremyLaurenson/octoprint-emc2101/blob/f27f5ec224c3f59dcdae5a213110e7c85edc94b1/octoprint_emc2101/tempcontrol.py#L9 will not work on anything other than Raspberry Pi OS, changed the username (not supported on OctoPi, but who's to say what they've done) or if someone is running multiple instances with the config directory changed. You could use an extra argument to the script to specify the path from the main plugin, collecting it with `sys.argv` in the script.

Below are some more nitpicky but still important comments because I like things to be as good as possible:
* https://github.com/JeremyLaurenson/octoprint-emc2101/blob/f27f5ec224c3f59dcdae5a213110e7c85edc94b1/octoprint_emc2101/__init__.py#L46-L48 and https://github.com/JeremyLaurenson/octoprint-emc2101/blob/f27f5ec224c3f59dcdae5a213110e7c85edc94b1/octoprint_emc2101/__init__.py#L89-L91 are not required, as `self.get_plugin_data_folder()` will ensure it exists.
* The rest of the settings code - why `conn.commit()` after every line? Would it not be quicker to execute all the lines in one go, and then write the changes? Bear in mind that `on_settings_save` is called synchronously with the settings API, so pressing save in the UI will slow down if it takes too long.

**Comment by JeremyLaurenson** [comment]
First, I appreciate all the feedback - its awesome to get more and I will continue to incorporate your ideas.

At the moment, the process is terminated when octoprint shuts down so there should be no issues with locking up resources, or restarting multiple unless octoprint itself has spun out of control and not called the callback... though the point is taken.

I feel like scheduling checks and doing the EMC2101 connections within the Octoprint framework has a hell of a lot more likelihood of hanging the all important printing process should something go wrong, like the sensor frying etc.
 

> Personally, I would just run your Python script in a thread - it would cut down the code needed in that file to around 1/3 of it's current length, because you can keep the settings in one place. But at the end of the day, as long as you don't break anything in OctoPrint that people will come complaining about, it is in your control 😉.

Will absolutely look into that. This .2 version is just the start. Whats the best method in your opinion, use  multiprocessing.Process? Id appreciate some pointers.

> 
> https://github.com/JeremyLaurenson/octoprint-emc2101/blob/f27f5ec224c3f59dcdae5a213110e7c85edc94b1/octoprint_emc2101/tempcontrol.py#L9 will not work on anything other than Raspberry Pi OS, changed the username (not supported on OctoPi, but who's to say what they've done) or if someone is running multiple instances with the config directory changed. You could use an extra argument to the script to specify the path from the main plugin, collecting it with `sys.argv` in the script.

Yup, will pass it args.. good catch.

> Below are some more nitpicky but still important comments because I like things to be as good as possible:
> 
> * https://github.com/JeremyLaurenson/octoprint-emc2101/blob/f27f5ec224c3f59dcdae5a213110e7c85edc94b1/octoprint_emc2101/__init__.py#L46-L48 and https://github.com/JeremyLaurenson/octoprint-emc2101/blob/f27f5ec224c3f59dcdae5a213110e7c85edc94b1/octoprint_emc2101/__init__.py#L89-L91 are not required, as `self.get_plugin_data_folder()` will ensure it exists.

Awesome... will remove.

> * The rest of the settings code - why `conn.commit()` after every line? Would it not be quicker to execute all the lines in one go, and then write the changes? Bear in mind that `on_settings_save` is called synchronously with the settings API, so pressing save in the UI will slow down if it takes too long.

Left over from debugging madness ;-)

**Comment by JeremyLaurenson** [comment]
Done remediating all the above. Keep 'em coming!

**Pull request closed without merging**